### ISSUE TITLE
Add support for AWS AppConfig

### ISSUE THREAD

**Issue by nerixim** [hashicorp/terraform-provider-aws#11973]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

> (snip) AppConfig, a new capability within AWS Systems Manager that makes it easy for customers to quickly roll out application configurations across applications hosted on EC2 instances, containers, Lambdas, mobile apps, IoT devices, and on-premise servers in a validated, controlled and monitored way. System administrators, DevOps engineers, and developers now have the ability to manage configuration changes, similar to the way they manage code, but without the need for deploying code or taking their application out of service when a configuration value changes, thus mitigating risk of potential outages.

### New or Affected Resource(s)

* aws_appconfig_application
* aws_appconfig_configuration_profile
* aws_appconfig_deployment_strategy
* aws_appconfig_environment
* aws_appconfig_deployment

### Potential Terraform Configuration

```hcl
resource "aws_appconfig_application" "this" {
  name = "sample"
  description = "just a sample"
  tags = {
    Name = "AppConfig Application"
  }
}

resource "aws_appconfig_configuration_profile" "this" {
  application_id = aws_appconfig_application.this.id
  name = "sample"
  description = "just a sample"
  location_uri = aws_ssm_parameter.this.arn | aws_ssm_document.this.arn | "https://ssm-url.aws"
  retrieval_role_arn = aws_iam_role.retrieve_ssm.arn
  validator {
    type = "JSON_SCHEMA" | "LAMBDA"
    content = "JSON Schema content or AWS Lambda function name"
  }
  tags = {
    Name = "AppConfig Configuration Profile"
  }
}

resource "aws_appconfig_deployment_strategy" "this" {
  name = "sample"
  description = "just a sample"
  deployment_duration_in_minutes = 15
  final_bake_time_in_minutes = 30
  growth_factor = 0.1
  growth_type = "LINEAR"
  replicate_to = "NONE" | "SSM_DOCUMENT"
  tags = {
    Name = "AppConfig Deployment Strategy"
  }
}

resource "aws_appconfig_environment" "this" {
  application_id = aws_appconfig_application.this.id
  name = "sample"
  description = "just a sample"
  monitors = [aws_cloudwatch_alarm.this.arn]
  tags = {
    Name = "AppConfig Environment"
  }
}

resource "aws_appconfig_deployment" "this" {
  application_id = aws_appconfig_application.this.id
  environment_id = aws_appconfig_environment.this.id
  deployment_strategy_id = aws_appconfig_deployment_strategy.this.id
  configuration_profile_id = aws_appconfig_configuration_profile.this.id
  configuration_version = "1"
  description = "just a sample"
  tags = {
    Name = "AppConfig Deployment"
  }
}
```

### References

* [Announcement](https://aws.amazon.com/about-aws/whats-new/2019/11/simplify-application-configuration-with-aws-appconfig/)
* [API Specs](https://docs.aws.amazon.com/appconfig/2019-10-09/APIReference/API_Operations.html)
* [ARN construction](https://docs.aws.amazon.com/service-authorization/latest/reference/list_awsappconfig.html#awsappconfig-resources-for-iam-policies)

**Comment by afonsoaugusto** [comment]
Hi,
Good morning,
any implementation forecast?

**Comment by LechuckThePirate** [comment]
Any news of when this will be available?

**Comment by hstefanov** [comment]
I am also waiting for this, but for now I am managing to workaround with a cloudformation stack as follows : 

resource "aws_cloudformation_stack" "appconfig" {
  name = "appconfig-stack"

  template_body = <<STACK
{
  "Parameters" : {
    "AppName" : {
      "Type" : "String",
      "Default" : "helloworld",
      "Description" : "Name of the application"
    },
    "EnvName" : {
      "Type" : "String",
      "Default" : "demo",
      "Description" : "Name of the environment"
    },
    "ProfileName" : {
      "Type" : "String",
      "Default" : "ExclamationPoints ",
      "Description" : "Name of the profile"
    }
  },
  "Resources" : {
    "appConfig": {
      "Type" : "AWS::AppConfig::Application",
      "Properties" : {
        "Name" : { "Ref" : "AppName" }
      }
    },
    "appConfigEnv": {
      "Type" : "AWS::AppConfig::Environment",
      "Properties" : {
        "Name" : { "Ref" : "EnvName" },
        "ApplicationId" : { "Ref" : "appConfig" }
      }
    },
    "appConfigProfile": {
      "Type" : "AWS::AppConfig::ConfigurationProfile",
      "Properties" : {
        "Name" : { "Ref" : "ProfileName" },
        "ApplicationId" : { "Ref" : "appConfig" }
      }
    }
  }
}
STACK
}

**Comment by LechuckThePirate** [comment]
Yes that's my approach as well but I think it would be good to add it

**Comment by larstobi** [comment]
SDK support is available since v1.25.42 https://github.com/aws/aws-sdk-go/pull/2977

**Comment by LechuckThePirate** [comment]
> SDK support is available since v1.25.42 [aws/aws-sdk-go#2977](https://github.com/aws/aws-sdk-go/pull/2977)

Nope, not as far as I know 

![image](https://user-images.githubusercontent.com/4067255/107748056-bf3caf00-6d18-11eb-8ada-f260e92a2b61.png)

**Comment by larstobi** [comment]
> Nope, not as far as I know

But, in that link that I posted, you'll see this:

> Release v1.25.42 (2019-11-25)
> Service Client Updates
> 
>    * service/alexaforbusiness: Updates service API and documentation
>    * service/appconfig: Adds new service

**Comment by marcomsousa** [comment]
The support in this provider isn't available, but the AWS SDK yes it is since 2019-11. That's means that to implement htis in this provider there is no blocker... just someone with time to implement this.

**Comment by shuheiktgw** [comment]
Let me see if I can implement this 👍

**Comment by deepshah22** [comment]
Superlike...!!!! Would love to see some progress here. This is super helpful.

**Comment by LechuckThePirate** [comment]
Which TF CLI version will have support for this?

**Comment by suzuki-shunsuke** [comment]
Hi, I have sent pull requests to support AppConfig.
To make pull request small, I split pull requests per resource.

#19307 Application + Environment
#19320 Configuration Profile
#19324 Hosted Configuration Version
#19359 Deployment Strategy

At first, please review #19307.

I think many users need this feature.

**Comment by anGie44** [comment]
Hi all, the following resources have been merged and will be released with `v3.50.0` of the TF AWS provider (likely out this Thursday):
* aws_appconfig_application
* aws_appconfig_configuration_profile
* aws_appconfig_deployment_strategy
* aws_appconfig_environment
* aws_appconfig_hosted_configuration_version

Big shoutout out to @suzuki-shunsuke and @Isaiah-Turner for providing the implementations!

**Comment by bryantbiggs** [comment]
this is awesome, thanks a ton! it looks like the only resource we might be missing is `aws_appconfig_deployment` or did I miss that somewhere in there?

**Comment by suzuki-shunsuke** [comment]
AS a prototype, I send a pull request to support aws_appconfig_deployment. #20172

AppConfig Deployment doesn't support common CRUD API.
Instead of common CRUD API, AppConfig Deployment supports the following API.

* get-deployment https://awscli.amazonaws.com/v2/documentation/api/latest/reference/appconfig/get-deployment.html
* start-deployment https://awscli.amazonaws.com/v2/documentation/api/latest/reference/appconfig/start-deployment.html
* stop-deployment https://awscli.amazonaws.com/v2/documentation/api/latest/reference/appconfig/stop-deployment.html

So I don't know how we should implement `Update` and `Delete`.

**Comment by anGie44** [comment]
Hi All, noting here that the last resource `aws_appconfig_deployment` has been merged and will be available in today's release of `v3.51.0` of the provider. Thanks again @suzuki-shunsuke for all your contributions 🎉
